### PR TITLE
3.0 beforeMarshal event

### PR DESCRIPTION
**Pull request by robertpustulka**
I like the idea of having `Model.beforeMarshal` event hinted in #5622 

Here is my take on this issue.

`$data` passed to marshaller is converted to `ArrayObject` for easy manipulation inside listeners and then casted back to array.

### CODE REVIEW

File: tests/TestCase/ORM/MarshallerTest.php [hunk 1]
```diff
@@ -1745,4 +1745,27 @@ public function testMergeWithCreate()
...
+    {
+        $data = [
+            'title' => 'My title',
+            'body' => 'My content'
+        ];
+
+        $marshall = new Marshaller($this->articles);
+
+        $this->articles->eventManager()->attach(function ($e, $data) {
+            $data['title'] = 'Modified title';
```
Comment (markstory) [inline tests/TestCase/ORM/MarshallerTest.php]: What about nested properties? It might be good to have a test for that as well.

**Comment by markstory** [comment]
Should there be only one event for the entire marshalling process or one for each entity that is created?

**Comment by dereuromark** [comment]
I bet rather for each entity, to assert the behaviors can modify it as designed/expected prior to running the validation on it.

**Comment by robertpustulka** [comment]
I think there should be one for each entity. So there would be consistent $data format in listeners. Also each model should prepare it's own related data for marshalling.

**Comment by markstory** [comment]
Ok. One each has a nice consistency with other orm events as well.

File: tests/TestCase/ORM/MarshallerTest.php (hunk 1 above)
Comment (robertpustulka) [inline tests/TestCase/ORM/MarshallerTest.php]: I'm wondering what should we do with nested properties. If beforeMarshal is fired for each created entity then should we be able to modify association data via parent's listeners?

Comment (markstory) [inline tests/TestCase/ORM/MarshallerTest.php]: I think allowing data modifications means you are stuck with earlier entities being able to modify the data for later ones. My original point was that this behavior should have assertions for it, as it is the kind of thing that could accidentally get broken.

Comment (robertpustulka) [inline tests/TestCase/ORM/MarshallerTest.php]: I'm not sure if i got it right :) but the test has been updated.

Comment (lorenzo) [inline tests/TestCase/ORM/MarshallerTest.php]: Basically what we wanted to see in the test is that the beforeMarshal event is triggered for the Users table as well

Comment (markstory) [inline tests/TestCase/ORM/MarshallerTest.php]: The added test looks good, but @lorenzo has a good point that coverage around related objects also getting events fired would be good.

Comment (robertpustulka) [inline tests/TestCase/ORM/MarshallerTest.php]: I've added assertions for associated entites

**Comment by markstory** [comment]
Latest changes look good. Nice work @robertpustulka

**Comment by lorenzo** [comment]
Thanks @robertpustulka !

**Pull request merged**